lib/common/http2client.c

```diff
@@ -297,16 +297,11 @@ static int on_head(struct st_h2o_http2client_conn_t *conn, struct st_h2o_http2cl
         stream->super._cb.on_head(&stream->super, is_end_stream ? h2o_httpclient_error_is_eos : NULL, 0, stream->input.status,
                                   h2o_iovec_init(NULL, 0), stream->input.headers.entries, stream->input.headers.size, (int)len, 0);
 
-    if (is_end_stream) {
+    if (is_end_stream || stream->super._cb.on_body == NULL) {
         close_stream(stream);
         return 0;
     }
 
-    if (stream->super._cb.on_body == NULL) {
-        ret = H2O_HTTP2_ERROR_PROTOCOL; // TODO: what error is suitable for this case?
```

Comment:
The answer depends on what we want to do.

If we are to allow applications to close the request by returning NULL in response to the `on_head` calleb with err (i.e. the second argument) set to NULL, we should send upstream a stream-level error and keep the connection alive. Otherwise, we would face errors on other requests sharing the same connection.

If we are to *not* allow applications to return NULL, we should do `assert(stream->supe._cb.on_body != NULL)`.

My assumption was that we are going for the latter approach, but I am definitely open to discussing what the correct way is, especially if there is a case that actually returns NULL.

Reply:
we're using NULL for some situations (e.g. redirect, websocket, etc) so we choose latter option here, and send RST_STREAM with CANCEL error code to the upstream.